handler/implementations/passThrough.go

```diff
@@ -136,97 +137,127 @@ func (h *PassThrough) Read(
-	}
-
 	var (
-		data string
-		ok   bool
-		err  error
+		sz  int
+		err error
 	)
 
+	logrus.Debugf("Executing Read() for req-id: %#x, handler: %s, resource: %s",
+		req.ID, h.Name, n.Name())
+
 	path := n.Path()
+	cntr := req.Container
+
 	prs := h.Service.ProcessService()
 	process := prs.ProcessCreate(req.Pid, req.Uid, req.Gid)
-	cntr := req.Container
 
 	//
-	// Caching here improves performance by avoiding dispatching the nsenter agent.  But
-	// note that caching is only helping processes at the sys container level, not in inner
-	// containers or unshared namespaces. To enable caching for those, we would need to
-	// have a cache per each namespace and this is expensive; plus we would also need to
-	// know when the namespace ceases to exist in order to destroy the cache associated
+	// The passthrough driver is slow because it must spawn a process that enters
+	// the container's namespaces (i.e., the nsenter agent) and read the data
+	// from there. To improve things, we cache the data on the first access to
+	// avoiding dispatching the nsenter agent on subsequent accesess.
+	//
+	// A couple of caveats on the caching:
+	//
+	// 1) Caching only done for processes at the sys container level, not in
+	// inner containers or inner unshared namespaces. To enable caching for
+	// those, we would need to have a cache per each namespace set (since the
+	// values under /proc/sys depend on the namespaces that the process belongs
+	// to). This would be expensive and would also require Sysbox to know when
+	// the namespace ceases to exist in order to destroy the cache associated
 	// with it.
 	//
+	// 2) As an optimization, we fetch data from the container's filesystem only
+	// when the req.Offset is 0. For req.Offset > 0, we assume that the data is
+	// be cached already. Without this optimization, we will likely go through
+	// fetchFile() twice for each read: one with req.Offset 0, and one at
+	// req.Offset X, where X is the number of bytes of the resource being
+	// read. That is, the handler's Read() method is normally invoked twice: the
+	// first read returns X bytes, the second read returns 0 bytes.
+
 	if domain.ProcessNsMatch(process, cntr.InitProc()) {
 
-		// If this resource is cached, return it's data; otherwise fetch its data from the
-		// host FS and store it in the cache.
 		cntr.Lock()
-		data, ok = cntr.Data(path, resource)
-		if !ok {
-			data, err = h.fetchFile(n, process)
+
+		// Check the data cache
+		sz, err = cntr.Data(path, req.Offset, &req.Data)
+		if err != nil && err != io.EOF {
+			cntr.Unlock()
+			return 0, fuse.IOerror{Code: syscall.EINVAL}
+		}
+
+		if req.Offset == 0 && sz == 0 && err == io.EOF {
+
+			// Resource is not cached, read it from the filesystem.
+			sz, err = h.fetchFile(process, n, req.Offset, &req.Data)
 			if err != nil {
 				cntr.Unlock()
-				return 0, err
+				return 0, fuse.IOerror{Code: syscall.EINVAL}
+			}
+
+			if sz == 0 {
+				cntr.Unlock()
+				return 0, nil
```

Comment:
No, in fact returning size = 0 denotes the end of file per the read() syscall man page:

      "On success, the number of bytes read is returned (zero indicates
       end of file), and the file position is advanced by this number.
       It is not an error if this number is smaller than the number of
       bytes requested; this may happen for example because fewer bytes
       are actually available right now (maybe because we were close to
       end-of-file, or because we are reading from a pipe, or from a
       terminal), or because read() was interrupted by a signal.  See
       also NOTES.

       On error, -1 is returned, and errno is set to indicate the error.
       In this case, it is left unspecified whether the file position
       (if any) changes."


So in fact we never want to pass io.EOF as an error to the syscall caller (i.e., an end-of-file is size = 0 with no error).

I verified this works properly.